spark/src/test/scala/org/apache/spark/sql/delta/CustomCatalogSuite.scala

```diff
@@ -0,0 +1,173 @@
+
+  test("RESTORE a table from DummyCatalog") {
+    val dummyCatalog =
+      spark.sessionState.catalogManager.catalog("dummy").asInstanceOf[DummyCatalog]
+    val tableName = "restore_table"
+    val tablePath = dummyCatalog.getTablePath(tableName)
+    withTable(tableName) {
+      sql("SET CATALOG dummy")
+      sql(f"CREATE TABLE $tableName (id bigint) USING delta")
+      sql("SET CATALOG spark_catalog")
+      // Insert some data into the table in the dummy catalog.
+      // To make it simple, here we insert data directly into the table path.
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (0)")
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (1)")
+      // Test 3-part identifier when the current catalog is the default catalog
+      sql(f"RESTORE TABLE dummy.default.$tableName VERSION AS OF 1")
+      checkAnswer(spark.table(f"dummy.default.$tableName"), spark.range(1).toDF())
+
+      sql("SET CATALOG dummy")
+      sql(f"RESTORE TABLE $tableName VERSION AS OF 0")
+      checkAnswer(spark.table(tableName), Nil)
+      sql(f"RESTORE TABLE $tableName VERSION AS OF 1")
+      checkAnswer(spark.table(tableName), spark.range(1).toDF())
+      // Test 3-part identifier
+      sql(f"RESTORE TABLE dummy.default.$tableName VERSION AS OF 2")
+      checkAnswer(spark.table(tableName), spark.range(2).toDF())
+      // Test file path table
+      sql(f"RESTORE TABLE delta.`$tablePath` VERSION AS OF 1")
+      checkAnswer(spark.table(tableName), spark.range(1).toDF())
+    }
+  }
+
+  test("Shallow Clone a table with time travel") {
+    val srcTable = "shallow_clone_src_table"
+    val destTable1 = "spark_catalog.default.shallow_clone_dest_table_1"
+    val destTable2 = "spark_catalog.default.shallow_clone_dest_table_2"
+    val destTable3 = "spark_catalog.default.shallow_clone_dest_table_3"
+    val destTable4 = "spark_catalog.default.shallow_clone_dest_table_4"
+    val dummyCatalog =
+      spark.sessionState.catalogManager.catalog("dummy").asInstanceOf[DummyCatalog]
+    val tablePath = dummyCatalog.getTablePath(srcTable)
+    withTable(srcTable, destTable1, destTable2, destTable3, destTable4) {
+      sql("SET CATALOG dummy")
+      sql(f"CREATE TABLE $srcTable (id bigint) USING delta")
+      sql("SET CATALOG spark_catalog")
+      // Insert some data into the table in the dummy catalog.
+      // To make it simple, here we insert data directly into the table path.
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (0)")
+      sql(f"INSERT INTO delta.`$tablePath` VALUES (1)")
+      // Test 3-part identifier when the current catalog is the default catalog
+      sql(f"CREATE TABLE $destTable1 SHALLOW CLONE dummy.default.$srcTable VERSION AS OF 1")
+      checkAnswer(spark.table(destTable1), spark.range(1).toDF())
+
+      sql("SET CATALOG dummy")
+      // Test simple shallow clone command under the dummy catalog
+      sql(f"CREATE TABLE $destTable2 SHALLOW CLONE $srcTable")
```

Comment:
Note: currently the target table to be created can be only from the session catalog.
https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/DeltaAnalysis.scala#L679

We should discuss how to fix this if the target is a table from an external DSV2 catalog.

cc @ryan-johnson-databricks

Reply:
Why is that, sorry? Is this just more legacy V1 code that can be removed as part of this PR? 
Or is there something more nefarious going on?

Reply:
```
        val finalTarget = new Path(statement.targetLocation.getOrElse(
          session.sessionState.catalog.defaultTablePath(tblIdent).toString))
```
To create a table with a clone table, it needs to have a target path. But if the catalog is an external catalog plugin, there is not such an API as `defaultTablePath`.  We can create a table and get the table path. However it is tricky to do the table creation during analysis.